routing/cross_mwm_index_graph.cpp

```diff
@@ -156,55 +160,66 @@ void CrossMwmIndexGraph::GetTwins(Segment const & s, bool isOutgoing, vector<Seg
   if (IsTransition(s, !isOutgoing))
     return;
 
   twins.clear();
-  // @TODO(bykoianko) It's necessary to check if mwm of |s| contains an A* cross mwm section
-  // and if so to use it. If not, osrm cross mwm sections should be used.
-
-  auto const getTwins = [&](TRoutingMappingPtr const & segMapping) {
-    vector<string> const & neighboringMwm = segMapping->m_crossContext.GetNeighboringMwmList();
 
-    for (string const & name : neighboringMwm)
-      InsertWholeMwmTransitionSegments(m_numMwmIds->GetId(CountryFile(name)));
-
-    auto it = m_transitionCache.find(s.GetMwmId());
-    CHECK(it != m_transitionCache.cend(), ());
-
-    vector<ms::LatLon> const & latLons =
-        isOutgoing ? GetLatLon(it->second.m_outgoing, s) : GetLatLon(it->second.m_ingoing, s);
-    for (string const & name : neighboringMwm)
-    {
-      NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(name));
-      auto const addFirstValidSegment = [&](TRoutingMappingPtr const & mapping) {
-        for (auto const & ll : latLons)
+  // Note. The code below looks for twins based on geometry index. This code works for
+  // any combination mwms with different cross mwm sections.
+  // It's possible to implement a faster version for two special cases:
+  // * all neighbouring mwms have cross_mwm section
+  // * all neighbouring mwms have osrm cross mwm sections
+  vector<m2::PointD> const transitions = GetTransitionPoints(s, isOutgoing);
+  for (m2::PointD const & p : transitions)
+  {
+    double constexpr kInvalidDistance = numeric_limits<double>::max();
+    bool exactConformityFound = false;
+    double minDistM = kInvalidDistance;
+    Segment minDistTwinSeg;
+    m_index.ForEachInRect([&](FeatureType & ft){
+      if (ft.GetID().m_mwmId.GetInfo()->GetType() != MwmInfo::COUNTRY)
+        return;
+
+      string const mwmName = ft.GetID().GetMwmName();
+      if (mwmName == FeatureID::kInvalidFileName)
+        return;
+
+      NumMwmId const numMwmId = m_numMwmIds->GetId(CountryFile(mwmName));
+      if (numMwmId == s.GetMwmId())
+        return;
+
+      ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      if (!m_vehicleModelFactory->GetVehicleModelForCountry(mwmName)->IsRoad(ft))
+        return;
+
+      vector<Segment> twinCandidates;
+      GetTransitions(ft, !isOutgoing, twinCandidates);
+      for (Segment const & tc : twinCandidates)
+      {
+        vector<m2::PointD> const twinPoints = GetTransitionPoints(tc, !isOutgoing);
+        for (m2::PointD const & tp : twinPoints)
         {
-          if (isOutgoing)
+          double const distM = MercatorBounds::DistanceOnEarth(p, tp);
+          if (distM == 0.0)
           {
-            mapping->m_crossContext.ForEachIngoingNodeNearPoint(
-                ll, [&](IngoingCrossNode const & node) {
-                  AddFirstValidSegment(mapping->m_segMapping, numMwmId, node.m_nodeId, twins);
-                });
+            twins.push_back(tc);
+            exactConformityFound = true;
```

Comment:
We need. Sometimes there're several segments which matches exactly. In that case all of them need to be added to |twins|. And they has to be added only once. This flag is used for this case.